File: octave/orthogonal_test.m
% TEST: linear system with orthogonal coefficient matrix

% 2x2 orthogonal matrix
A = [ 1  1;
      1 -1 ];
A = A*1.0/sqrt(2.0)
% Known terms vector
b = [ 2; 3 ]
% Solve the system (skip orthogonality check)
x = orthogonal(A, b, 0)
% Check
if allclose(b, A*x, 0.0001) == 0
    error('Orthogonal test failure')
end

% 3x3 orthogonal matrix
A = [ 2 -2  1;
      1  2  2;
      2  1 -2 ];
A = A*1.0/3.0
% Known terms vector
b = [ 2; 3; 4 ]
% Solve the system (with orthogonality check)
x = orthogonal(A, b, 1)
% Check
if allclose(b, A*x, eps) == 0
    error('Error: orthogonal test failure')
end

File: python/orthogonal_test.py
# Orthogonal linear system solver tests

import numpy as np
from math import sqrt
from orthogonal import orthogonal

################################################################################

# 2x2 orthogonal matrix
A = np.matrix('1  1;'
              '1 -1', float)
A = A*1.0/sqrt(2.0)
# Known terms vector
b = np.matrix('2; 3')
# Solve the system
x = orthogonal(A, b, 1)
# Check
if np.allclose(b, A*x) == False:
    raise Exception('Orthogonal test failure')

################################################################################

# 2x2 orthogonal matrix
A = np.matrix('2 -2  1;'
              '1  2  2;'
              '2  1 -2', float)
A = A*1.0/3.0
# Known terms vector
b = np.matrix('2; 3; 4')
# Solve the system
x = orthogonal(A, b)
# Check
if np.allclose(b, A*x) == False:
    raise Exception('Orthogonal test failure')